core/kernel/shared/sprintf.rb

```diff
@@ -0,0 +1,945 @@
+      # from docs - Convert floating point argument as [-]0xh.hhhhp[+-]dd
+      # * p+7 => p+07 - two digits for exponenta
+      # * fraction part should be 4 digits long by default
+      it "converts floating point argument as [-]0xh.hhhhp[+-]dd" do
+        format("%a", 196).should == "0x1.88p+7"
+        format("%a", -196).should == "-0x1.88p+7"
+        format("%a", 196.1).should == "0x1.8833333333333p+7"
+        format("%a", 0.01).should == "0x1.47ae147ae147bp-7"
+        format("%a", -0.01).should == "-0x1.47ae147ae147bp-7"
+      end
+
+      it "displays Float::INFINITY as Inf" do
+        format("%a", Float::INFINITY).should == "Inf"
+        format("%a", -Float::INFINITY).should == "-Inf"
+      end
+
+      it "displays Float::NAN as NaN" do
+        format("%a", Float::NAN).should == "NaN"
+        format("%a", -Float::NAN).should == "NaN"
+      end
+    end
+
+    describe "A" do
+      it "converts floating point argument as [-]0xh.hhhhp[+-]dd and use uppercase X and P" do
+        format("%A", 196).should == "0X1.88P+7"
+        format("%A", -196).should == "-0X1.88P+7"
+        format("%A", 196.1).should == "0X1.8833333333333P+7"
+        format("%A", 0.01).should == "0X1.47AE147AE147BP-7"
+        format("%A", -0.01).should == "-0X1.47AE147AE147BP-7"
+      end
+
+      it "displays Float::INFINITY as Inf" do
+        format("%A", Float::INFINITY).should == "Inf"
+        format("%A", -Float::INFINITY).should == "-Inf"
+      end
+
+      it "displays Float::NAN as NaN" do
+        format("%A", Float::NAN).should == "NaN"
+        format("%A", -Float::NAN).should == "NaN"
+      end
+    end
+  end
+
+  describe "other formats" do
+    describe "c" do
+      it "displays character if argument is a numeric code of character" do
+        format("%c", 97).should == "a"
+      end
+
+      it "displays character if argument is a single character string" do
+        format("%c", "a").should == "a"
+      end
+
+      it "raises ArgumentError if argument is a string of several characters" do
+        -> () {
+          format("%c", "abc")
+        }.should raise_error(ArgumentError)
+      end
+
+      it "supports Unicode characters" do
+        format("%c", 1286).should == "Ԇ"
+        format("%c", "ش").should == "ش"
+      end
+    end
+
+    describe "p" do
+      it "displays argument.inspect value" do
+        obj = mock("object")
+        obj.should_receive(:inspect).and_return("<inspect-result>")
+        format("%p", obj).should == "<inspect-result>"
+      end
+    end
+
+    describe "s" do
+      it "substitute argument passes as a string" do
+        format("%s", "abc").should == "abc"
+      end
+
+      it "converts argument to string with to_s" do
+        obj = mock("string")
+        obj.should_receive(:to_s).and_return("abc")
+        format("%s", obj).should == "abc"
+      end
+
+      # strange behavior
+      # expected any core/std lib uses to_int/to_str/to_ary/to_hash for argument
```

Comment:
So it uses `to_s` instead? Could you add a spec for that?
I think core APIs are not all consistent with this, particularly for `to_str`/`to_s`.
Usually, only String-like values should implement `to_str`, while `to_s` is meant to convert an object to a String representation, which makes sense here.

Reply:
TLDNR: Agreed

`to_str` is used for implicit coercion, i.e. any core method that accepts a String should accept an object responding to `to_str` (e.g. `define_method`).

`to_s` is used for explicit coercion. Most core methods that expect a String will not call `to_s` on their argument and instead raise a TypeError. The few exceptions are methods that are designed to convert to strings, like `"#{foo}"`, or `"%{x}" % {x: foo}`, or here `sprintf`.

It is expected that `to_s` can be lossy, i.e. it gives you a string output from a more complex object. `to_str` is reserved for cases where the object "almost is a" string. In particular, it is expected that any object responding to `to_str` will also respond to `to_s` and both should give the same result.